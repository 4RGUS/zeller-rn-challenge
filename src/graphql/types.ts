export interface TableStringFilterInput {
  ne?: string;
  eq?: string;
  le?: string;
  lt?: string;
  ge?: string;
  gt?: string;
  contains?: string;
  notContains?: string;
  between?: string[];
  beginsWith?: string;
}

export interface TableZellerCustomerFilterInput {
  id?: TableStringFilterInput;
  name?: TableStringFilterInput;
  email?: TableStringFilterInput;
  role?: TableStringFilterInput;
}

export interface ZellerCustomer {
  id: string;
  name?: string;
  email?: string;
  role?: string;
}

export interface ZellerCustomerConnection {
  items?: ZellerCustomer[];
  nextToken?: string;
}

export interface ListZellerCustomersQueryVariables {
  filter?: TableZellerCustomerFilterInput;
  limit?: number;
  nextToken?: string;
}

export interface ListZellerCustomersQuery {
  listZellerCustomers?: ZellerCustomerConnection;
}

export interface GetZellerCustomerQueryVariables {
  id: string;
}

export interface GetZellerCustomerQuery {
  getZellerCustomer?: ZellerCustomer;
}

export enum Roles {
  ADMIN = "ADMIN",
  MANAGER = "MANAGER",
}
