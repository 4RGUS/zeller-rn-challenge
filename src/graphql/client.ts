import {
  ApolloClient,
  ApolloLink,
  HttpLink,
  InMemoryCache,
} from "@apollo/client";
import awsconfig from "./aws-exports";
import { AuthOptions, createAuthLink } from "aws-appsync-auth-link";
import { createSubscriptionHandshakeLink } from "aws-appsync-subscription-link";

const auth: AuthOptions = {
  type: "API_KEY",
  apiKey: awsconfig.aws_appsync_apiKey,
};

const httpLink = new HttpLink({
  uri: awsconfig.aws_appsync_graphqlEndpoint,
});

const link = ApolloLink.from([
  createAuthLink({
    url: awsconfig.aws_appsync_graphqlEndpoint,
    region: awsconfig.aws_appsync_region,
    auth,
  }),
  createSubscriptionHandshakeLink(
    {
      url: awsconfig.aws_appsync_graphqlEndpoint,
      region: awsconfig.aws_appsync_region,
      auth,
    },
    httpLink
  ),
]);

const client = new ApolloClient({
  link,
  cache: new InMemoryCache(),
});

export default client;
