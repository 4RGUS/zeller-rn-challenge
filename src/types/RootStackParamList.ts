import { NativeStackNavigationProp } from "@react-navigation/native-stack";

export type RootStackParamList = {
  Home: undefined;
  UserList: undefined;
};

export type UserListScreenNavigationProp = NativeStackNavigationProp<
  RootStackParamList,
  "UserList"
>;
