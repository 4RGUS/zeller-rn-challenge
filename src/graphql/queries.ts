import { gql } from "@apollo/client";

export const LIST_ZELLER_CUSTOMERS = gql`
  query ListZellerCustomers($filter: TableZellerCustomerFilterInput) {
    listZellerCustomers(filter: $filter) {
      items {
        id
        name
        email
        role
      }
      nextToken
    }
  }
`;

export const SEARCH_ZELLER_CUSTOMERS = gql`
  query SearchZellerCustomers(
    $nameFilter: TableStringFilterInput
    $limit: Int
    $nextToken: String
  ) {
    listZellerCustomers(
      filter: { name: $nameFilter }
      limit: $limit
      nextToken: $nextToken
    ) {
      items {
        id
        name
        email
        role
      }
      nextToken
    }
  }
`;
