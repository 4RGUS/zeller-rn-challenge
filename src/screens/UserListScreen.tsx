import React, { useCallback, useState } from "react";
import { View, Text, ActivityIndicator } from "react-native";
import { useQuery } from "@apollo/client";
import { LIST_ZELLER_CUSTOMERS } from "../graphql/queries";
import { Listing } from "../components/Listing";
import { Roles } from "../graphql/types";
import { UserTypeButtons } from "../components/UserTypeButtons";

const UserListScreen = () => {
  const [selectedRole, setSelectedRole] = useState(Roles.ADMIN);

  const { data, loading, error, refetch } = useQuery(LIST_ZELLER_CUSTOMERS, {
    variables: {
      filter: {
        role: {
          eq: selectedRole,
        },
      },
    },
  });

  const onPressHandle = useCallback(
    (role: Roles) => {
      setSelectedRole(role);
      refetch({
        filter: {
          role: {
            eq: role,
          },
        },
      });
    },
    [setSelectedRole, refetch]
  );

  if (loading)
    return (
      <ActivityIndicator testID="loading-spinner" size="large" color="#000" />
    );

  return (
    <View style={{ flex: 1, padding: 30 }}>
      <Text style={{ fontSize: 20, marginBottom: 16 }}>User Types</Text>
      <UserTypeButtons onPress={onPressHandle} selectedRole={selectedRole} />

      <Text style={{ fontSize: 20, marginVertical: 16 }}>
        {selectedRole} Users
      </Text>

      <View style={{ flex: 1 }}>
        {data ? (
          <Listing
            roleType={selectedRole}
            data={data?.listZellerCustomers.items}
          />
        ) : null}
        {error ? (
          <Text testID="error-message">Error: {error.message}</Text>
        ) : null}
      </View>
    </View>
  );
};

export default UserListScreen;
