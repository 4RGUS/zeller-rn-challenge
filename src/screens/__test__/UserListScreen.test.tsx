import React from "react";
import { render } from "@testing-library/react-native";
import { MockedProvider } from "@apollo/client/testing";
import UserListScreen from "../UserListScreen"; // adjust path if needed
import { LIST_ZELLER_CUSTOMERS } from "../../graphql/queries";
import { Roles } from "../../graphql/types";

describe("UserListScreen Snapshot", () => {
  it("matches snapshot when loading", () => {
    const { toJSON } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <UserListScreen />
      </MockedProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("matches snapshot with data", async () => {
    const mocks = [
      {
        request: {
          query: LIST_ZELLER_CUSTOMERS,
          variables: {
            filter: { role: { eq: Roles.ADMIN } },
          },
        },
        result: {
          data: {
            listZellerCustomers: {
              items: [
                { id: "1", name: "John Doe", role: Roles.ADMIN },
                { id: "2", name: "Jane Smith", role: Roles.ADMIN },
              ],
            },
          },
        },
      },
    ];

    const { toJSON, findByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <UserListScreen />
      </MockedProvider>
    );

    await findByText("John Doe");
    expect(toJSON()).toMatchSnapshot();
  });

  it("matches snapshot on error", async () => {
    const errorMocks = [
      {
        request: {
          query: LIST_ZELLER_CUSTOMERS,
          variables: {
            filter: { role: { eq: Roles.ADMIN } },
          },
        },
        error: new Error("Something went wrong!"),
      },
    ];

    const { toJSON, findByTestId } = render(
      <MockedProvider mocks={errorMocks} addTypename={false}>
        <UserListScreen />
      </MockedProvider>
    );

    await findByTestId("error-message");
    expect(toJSON()).toMatchSnapshot();
  });
});
