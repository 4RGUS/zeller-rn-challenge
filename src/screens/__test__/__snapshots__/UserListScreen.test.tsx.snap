// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListScreen Snapshot matches snapshot on error 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 30,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 20,
        "marginBottom": 16,
      }
    }
  >
    User Types
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#E6F0FF",
        "borderRadius": 8,
        "flexDirection": "row",
        "gap": 12,
        "marginBottom": 8,
        "opacity": 1,
        "padding": 12,
        "width": "100%",
      }
    }
    testID="selected-role"
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#4D94FF",
          "borderRadius": 9,
          "borderWidth": 2,
          "height": 18,
          "justifyContent": "center",
          "width": 18,
        }
      }
      testID="outer-circle"
    >
      <View
        style={
          {
            "backgroundColor": "#4D94FF",
            "borderRadius": 4,
            "height": 8,
            "width": 8,
          }
        }
        testID="inner-circle"
      />
    </View>
    <Text>
      Admin
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "borderRadius": 8,
        "flexDirection": "row",
        "gap": 12,
        "marginBottom": 8,
        "opacity": 1,
        "padding": 12,
        "width": "100%",
      }
    }
    testID="un-selected-role"
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#4D94FF",
          "borderRadius": 9,
          "borderWidth": 2,
          "height": 18,
          "justifyContent": "center",
          "width": 18,
        }
      }
      testID="outer-circle"
    />
    <Text>
      Manager
    </Text>
  </View>
  <Text
    style={
      {
        "fontSize": 20,
        "marginVertical": 16,
      }
    }
  >
    Admin
     Users
  </Text>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      testID="error-message"
    >
      Error: 
      Something went wrong!
    </Text>
  </View>
</View>
`;

exports[`UserListScreen Snapshot matches snapshot when loading 1`] = `
<ActivityIndicator
  color="#000"
  size="large"
  testID="loading-spinner"
/>
`;

exports[`UserListScreen Snapshot matches snapshot with data 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 30,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 20,
        "marginBottom": 16,
      }
    }
  >
    User Types
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#E6F0FF",
        "borderRadius": 8,
        "flexDirection": "row",
        "gap": 12,
        "marginBottom": 8,
        "opacity": 1,
        "padding": 12,
        "width": "100%",
      }
    }
    testID="selected-role"
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#4D94FF",
          "borderRadius": 9,
          "borderWidth": 2,
          "height": 18,
          "justifyContent": "center",
          "width": 18,
        }
      }
      testID="outer-circle"
    >
      <View
        style={
          {
            "backgroundColor": "#4D94FF",
            "borderRadius": 4,
            "height": 8,
            "width": 8,
          }
        }
        testID="inner-circle"
      />
    </View>
    <Text>
      Admin
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFF",
        "borderRadius": 8,
        "flexDirection": "row",
        "gap": 12,
        "marginBottom": 8,
        "opacity": 1,
        "padding": 12,
        "width": "100%",
      }
    }
    testID="un-selected-role"
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#4D94FF",
          "borderRadius": 9,
          "borderWidth": 2,
          "height": 18,
          "justifyContent": "center",
          "width": 18,
        }
      }
      testID="outer-circle"
    />
    <Text>
      Manager
    </Text>
  </View>
  <Text
    style={
      {
        "fontSize": 20,
        "marginVertical": 16,
      }
    }
  >
    Admin
     Users
  </Text>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      data={
        [
          {
            "id": "1",
            "name": "John Doe",
            "role": "Admin",
          },
          {
            "id": "2",
            "name": "Jane Smith",
            "role": "Admin",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "gap": 12,
                "marginBottom": 8,
                "padding": 12,
                "paddingLeft": 0,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#E6F0FF",
                  "borderRadius": 5,
                  "height": 45,
                  "justifyContent": "center",
                  "width": 45,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#4D94FF",
                    "fontSize": 18,
                  }
                }
              >
                J
              </Text>
            </View>
            <View
              style={
                {
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "fontWeight": "bold",
                  }
                }
              >
                John Doe
              </Text>
              <Text
                style={
                  {
                    "color": "#A9A9A9",
                  }
                }
              >
                Admin
              </Text>
            </View>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "gap": 12,
                "marginBottom": 8,
                "padding": 12,
                "paddingLeft": 0,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#E6F0FF",
                  "borderRadius": 5,
                  "height": 45,
                  "justifyContent": "center",
                  "width": 45,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#4D94FF",
                    "fontSize": 18,
                  }
                }
              >
                J
              </Text>
            </View>
            <View
              style={
                {
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "fontWeight": "bold",
                  }
                }
              >
                Jane Smith
              </Text>
              <Text
                style={
                  {
                    "color": "#A9A9A9",
                  }
                }
              >
                Admin
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
